jwst/assign_wcs/miri.py

```diff
@@ -134,102 +136,212 @@ def imaging_distortion(input_model, reference_files):
+    # And compute the v2,v3 coordinates of pixels in this reference row
+    v2v3refrow = np.array(subarray_dist(xrow, yrow))[:, 0, :]
+
+    # Now repeat the v2,v3, matrix from the central row so that it is copied to all of the other valid rows too
+    v2_full = mb.repmat(v2v3refrow[0], _toindex(bb[1][1]) + 1 - _toindex(bb[1][0]), 1)
+    v3_full = mb.repmat(v2v3refrow[1], _toindex(bb[1][1]) + 1 - _toindex(bb[1][0]), 1)
+    # v2_full and v3_full now have shape 392x68 for slitless and 391x44 for slit
+
+    # Now take these matrices and put them into tabular models that can be interpolated to find v2,v3 for arbitrary
+    # x,y pixel values in the valid region.
+    v2_t2d = models.Tabular2D(lookup_table=v2_full, name='v2table',
+                              bounds_error=False, fill_value=np.nan)
+    v3_t2d = models.Tabular2D(lookup_table=v3_full, name='v3table',
+                              bounds_error=False, fill_value=np.nan)
+    
+    # Now deal with the fact that the spectral trace isn't perfectly up and down along detector.
+    # This information is contained in the xcenter/ycenter values in the CDP table, but we'll handle it
+    # as a simple rotation using a linear fit to this relation provided by the CDP.
+
+    z=np.polyfit(xcen,ycen,1)
+    slope=1./z[0]
+    traceangle = np.arctan(slope)*180./np.pi # trace angle in degrees
+    rot = models.Rotation2D(traceangle) # Rotation model   
+
+    # Now include this rotation in our overall transform
+    # First shift to a frame relative to the trace zeropoint, then apply the rotation
+    # to correct for the curved trace.  End in a rotated frame relative to zero at the reference point
+    # and where yrot is aligned with the spectral trace)
+    xysubtoxyrot=models.Shift(-zero_point[0]) & models.Shift(-zero_point[1]) | rot
+    # Next shift back to the subarray frame, and then map to v2v3
+    xyrottov2v3=models.Shift(zero_point[0]) & models.Shift(zero_point[1]) | models.Mapping((1, 0, 1, 0)) | v2_t2d & v3_t2d
+    # The two models together
+    xysubtov2v3=xysubtoxyrot | xyrottov2v3
+
+    # Work out the spectral component of the transform
+    # First compute the reference trace in the rotated-Y frame
+    xcenrot,ycenrot=rot(xcen,ycen)
+    # The input table of wavelengths isn't perfect, and the delta-wavelength steps show some unphysical behaviour
+    # Therefore fit with a spline for the ycenrot->wavelength transform
+    # Reverse vectors so that yinv is increasing (needed for spline fitting function)
+    yrev=ycenrot[::-1]
+    wrev=wavetab[::-1]
+    # Spline fit with enforced smoothness
+    spl=UnivariateSpline(yrev,wrev,s=0.002)
+    # Evaluate the fit at the rotated-y reference points
+    wavereference=spl(yrev)
+    # wavereference now contains the wavelengths corresponding to regularly-sampled ycenrot, create the model
+    wavemodel = models.Tabular1D(lookup_table=wavereference, points=yrev,name='waveref',bounds_error=False, fill_value=np.nan)
```

Comment:
At some point would it be desirable to drop the `Tabular1D` model and use the spline directly? It will require writing the serialization code for the spline.

Reply:
Although I ran a quick test and I got slightly different results when using the spline instead of the Tabular model (I am guessing because of the interpolation) but that may not be significant.

Reply:
Sure, this could be a long-term thing to do.  I think I used Tabular1D since I wasn't sure how to do the interpolation to arbitrary locations directly from the spline model, but doing it directly would probably be preferable.  I imagine that any small differences probably result from the Tabular1D linear interpolation between nearest points just approximating to the actual spline fit.